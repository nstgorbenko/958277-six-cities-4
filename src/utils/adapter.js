export const adaptOffer = (offer) => ({
  id: offer.id,
  city: {
    name: offer.city.name,
    location: {
      coordinates: [offer.city.location.latitude, offer.city.location.longitude],
      zoom: offer.city.location.zoom,
    },
  },
  location: {
    coordinates: [offer.location.latitude, offer.location.longitude],
    zoom: offer.location.zoom,
  },
  name: offer.title,
  type: offer.type,
  description: offer.description,
  price: offer.price,
  photo: offer.preview_image,
  allPhotos: offer.images,
  bedrooms: offer.bedrooms,
  adults: offer.max_adults,
  amenities: offer.goods,
  host: {
    id: offer.host.id,
    name: offer.host.name,
    avatar: offer.host.avatar_url,
    isSuper: offer.host.is_pro,
  },
  rating: offer.rating,
  isPremium: offer.is_premium,
  isFavorite: offer.is_favorite,
});

export const adaptOffers = (offers) =>
  offers.map((offer) => adaptOffer(offer));

export const adaptUserInfo = (info) => ({
  id: info.id,
  name: info.name,
  email: info.email,
  avatar: info.avatar_url,
  isSuper: info.is_pro,
});
