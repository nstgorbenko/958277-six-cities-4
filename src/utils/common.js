export const capitalizeWord = (word) => word[0].toUpperCase() + word.slice(1);

export const getFirstCity = (offers) => {
  const {name, location} = offers[0];
  return ({name, location});
};

export const groupOffersByCities = (offers) => {
  const allCities = offers.reduce((cities, offer) => {
    const currentCity = offer.city.name;
    if (!cities.hasOwnProperty(currentCity)) {
      cities[currentCity] = {
        name: offer.city.name,
        location: offer.city.location,
        offers: [],
      };
    }
    cities[currentCity].offers.push(offer);

    return cities;
  }, {});

  return Object.values(allCities);
};

export const getRatingPercent = (rating) => {
  return Math.round(rating) * 20;
};

export const noop = () => {
  // do nothing
};

export const updateNearbyOffers = (oldNearbyOffers, newNearbyOffer) => {

  if (oldNearbyOffers.length > 0) {
    const newNearbyOfferIndex = oldNearbyOffers.findIndex(({id}) => id === newNearbyOffer.id);
    oldNearbyOffers[newNearbyOfferIndex] = newNearbyOffer;
  }

  return oldNearbyOffers;
};

export const updateOffers = (oldOffers, newOffer) => {
  const city = newOffer.city.name;

  if (oldOffers.length > 0) {
    const cityOffers = oldOffers.filter(({name}) => name === city)[0].offers;
    const newOfferIndex = cityOffers.findIndex(({id}) => id === newOffer.id);
    cityOffers[newOfferIndex] = newOffer;

    return oldOffers;
  }

  return groupOffersByCities([newOffer]);
};

export const updateFavorites = (oldFavorites, newFavorite) => {
  const isFavorite = newFavorite.isFavorite;

  if (isFavorite) {
    oldFavorites.push(newFavorite);
  } else {
    const index = oldFavorites.findIndex(({id}) => id === newFavorite.id);
    oldFavorites.splice(index, 1);
  }

  return oldFavorites;
};
