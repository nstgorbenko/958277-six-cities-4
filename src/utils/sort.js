import {SortType} from "../const.js";

const sortEventsByPriceToHigh = (a, b) => a.price - b.price;

const sortEventsByPriceToLow = (a, b) => b.price - a.price;

const sortEventsByRating = (a, b) => b.rating - a.rating;

export const getSortedPlaces = ([...places], sortType) => {
  switch (sortType) {
    case SortType.POPULAR:
      return places;
    case SortType.PRICE_TO_HIGH:
      return places.sort(sortEventsByPriceToHigh);
    case SortType.PRICE_TO_LOW:
      return places.sort(sortEventsByPriceToLow);
    case SortType.TOP_RATED:
      return places.sort(sortEventsByRating);
    default:
      throw new Error(`Unknown sort type: ${sortType}`);
  }
};
