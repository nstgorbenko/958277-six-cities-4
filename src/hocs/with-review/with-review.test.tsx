import * as React from "react";
import * as renderer from "react-test-renderer";

import {noop} from "../../utils/common";
import withReview from "./with-review";

const MockComponent = () => <div />;

describe(`withReview HOC rendering`, () => {
  it(`renders correctly`, () => {
    const MockComponentWrapped = withReview(MockComponent);

    const tree = renderer
      .create(
          <MockComponentWrapped
            id={1}
            rating={0}
            review={``}
            onChange={noop}
            onReset={noop}
          />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
