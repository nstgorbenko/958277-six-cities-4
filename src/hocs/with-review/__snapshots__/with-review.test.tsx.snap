// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withReview HOC rendering renders correctly 1`] = `<div />`;
